pkg/vault/contracts/authorizer/TimelockAuthorizerMigrator.sol

```diff
@@ -0,0 +1,166 @@
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/math/Math.sol";
+
+import "./IBasicAuthorizer.sol";
+import "./TimelockAuthorizer.sol";
+import "../interfaces/IVault.sol";
+
+contract TimelockAuthorizerMigrator {
+    address public constant EVERYWHERE = address(-1);
+    bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+
+    // solhint-disable var-name-mixedcase
+    bytes32 public immutable GRANT_PERMISSION_ACTION_ID;
+    bytes32 public immutable REVOKE_PERMISSION_ACTION_ID;
+
+    IVault public immutable vault;
+    IBasicAuthorizer public immutable oldAuthorizer;
+    TimelockAuthorizer public immutable newAuthorizer;
+
+    uint256 public migratedRoles;
+    OldRoleData[] public rolesData;
+
+    /**
+     * @dev This structure is required to tell the migrator which roles need to be
+     * granted for which targets in the new timelock authorizer.
+     */
+    struct OldRoleData {
+        bytes32 role;
+        address target;
+    }
+
+    constructor(
+        IVault _vault,
+        IBasicAuthorizer _oldAuthorizer,
+        OldRoleData[] memory _rolesData
+    ) {
+        TimelockAuthorizer _newAuthorizer = new TimelockAuthorizer(address(this));
+        newAuthorizer = _newAuthorizer;
+        oldAuthorizer = _oldAuthorizer;
+        vault = _vault;
+
+        for (uint256 i = 0; i < _rolesData.length; i++) {
+            rolesData.push(OldRoleData(_rolesData[i].role, _rolesData[i].target));
+        }
+
+        bytes32 id = bytes32(uint256(address(_newAuthorizer)));
+        GRANT_PERMISSION_ACTION_ID = keccak256(abi.encodePacked(id, TimelockAuthorizer.grantPermissions.selector));
+        REVOKE_PERMISSION_ACTION_ID = keccak256(abi.encodePacked(id, TimelockAuthorizer.revokePermissions.selector));
+    }
+
+    /**
+     * @dev Migrate roles from the old authorizer to the new one
+     * @param n Number of roles to migrate, use 0 to migrate all the remaining ones
+     */
+    function migrate(uint256 n) external {
+        _beforeMigrate();
+        _migrate(n == 0 ? rolesData.length : n);
+        _afterMigrate();
+    }
+
+    function _migrate(uint256 n) internal {
+        uint256 i = migratedRoles;
+        uint256 to = Math.min(i + n, rolesData.length);
+        for (; i < to; i++) _migrate(rolesData[i]);
+        migratedRoles = i;
+    }
+
+    function _migrate(OldRoleData memory roleData) internal {
+        // In the old authorizer, each role has an admin role which might have been granted to many admins.
+        // In order to replicate that in the new authorizer, we need to grant permission for
+        // `TimelockAuthorizer.grantPermissions` to the same list of admins on each target.
+        // First, the contract iterates over the accounts that had the role granted in the old authorizer, granting
+        // the permission for the same role for the specified target in the new authorizer.
```

Comment:
This reads a bit weird as the first paragraph is unrelated to the second one, and to the first `_migrate` call. I'd move it below.